libpod/runtime_ctr.go

```diff
@@ -915,6 +915,44 @@ func (r *Runtime) evictContainer(ctx context.Context, idOrName string, removeVol
 	return id, cleanupErr
 }
 
+// DependMap populates of map of all containers depending on the specified container
+func (r *Runtime) DependMap(ctx context.Context, ctr *Container, ctrMap map[string]*Container) error {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	// Grab all containers
+	allctrs, err := r.GetAllContainers()
+	if err != nil {
+		return err
+	}
+	// Make sure current Container is mapped
+	if ctrMap[ctr.ID()] == nil {
+		ctrMap[ctr.ID()] = ctr
+	}
+	return r.dependMap(ctx, ctr, allctrs, ctrMap)
+}
+
+// dependMap populates of map of all containers depending on the specified container
+func (r *Runtime) dependMap(ctx context.Context, searchCtr *Container, allCtrs []*Container, ctrMap map[string]*Container) error {
+	for _, ctr := range allCtrs {
+		if ctr == searchCtr {
+			continue
+		}
+		if ctrMap[ctr.ID()] != nil {
+			continue
+		}
+		for _, cid := range ctr.Dependencies() {
+			if cid == searchCtr.ID() {
+				ctrMap[ctr.ID()] = ctr
```

Comment:
Can we turn the map into `map[string]string`?

Currently there is a bug since we're using the loop variable. `ctr` will be added to the map but will change with the next iteration of the loop. That means all entries of `ctr` will point to the last item in `allCtrs`. Using a `map[string]string` will avoid that problem _AND_ it forces callers to look up the containers another time before removal - they may have already been removed in the meantime.

Reply:
Rewritten.